main.tf

```diff
@@ -78,41 +167,41 @@ resource "aws_codebuild_project" "default" {
     type = "CODEPIPELINE"
   }
 
+  # The cache as a list with a map object inside.
+  cache = ["${local.cache}"]
+
   environment {
     compute_type    = "${var.build_compute_type}"
     image           = "${var.build_image}"
     type            = "LINUX_CONTAINER"
     privileged_mode = "${var.privileged_mode}"
 
-    environment_variable {
+    environment_variable = [{
       "name"  = "AWS_REGION"
       "value" = "${signum(length(var.aws_region)) == 1 ? var.aws_region : data.aws_region.default.name}"
-    }
-
-    environment_variable {
-      "name"  = "AWS_ACCOUNT_ID"
-      "value" = "${signum(length(var.aws_account_id)) == 1 ? var.aws_account_id : data.aws_caller_identity.default.account_id}"
-    }
-
-    environment_variable {
-      "name"  = "IMAGE_REPO_NAME"
-      "value" = "${signum(length(var.image_repo_name)) == 1 ? var.image_repo_name : "UNSET"}"
-    }
-
-    environment_variable {
-      "name"  = "IMAGE_TAG"
-      "value" = "${signum(length(var.image_tag)) == 1 ? var.image_tag : "latest"}"
-    }
-
-    environment_variable {
-      "name"  = "STAGE"
-      "value" = "${var.stage}"
-    }
-
-    environment_variable {
-      "name"  = "GITHUB_TOKEN"
-      "value" = "${var.github_token}"
-    }
+    },
+      {
+        "name"  = "AWS_ACCOUNT_ID"
+        "value" = "${signum(length(var.aws_account_id)) == 1 ? var.aws_account_id : data.aws_caller_identity.default.account_id}"
+      },
+      {
+        "name"  = "IMAGE_REPO_NAME"
+        "value" = "${signum(length(var.image_repo_name)) == 1 ? var.image_repo_name : "UNSET"}"
+      },
+      {
+        "name"  = "IMAGE_TAG"
+        "value" = "${signum(length(var.image_tag)) == 1 ? var.image_tag : "latest"}"
+      },
+      {
+        "name"  = "STAGE"
+        "value" = "${signum(length(var.stage)) == 1 ? var.stage : "UNSET"}"
+      },
+      {
+        "name"  = "GITHUB_TOKEN"
+        "value" = "${signum(length(var.github_token)) == 1 ? var.github_token : "UNSET"}"
+      },
+      "${var.environment_variables}",
```

Comment:
This was part of the issue, if I parsed that list of maps through a interpolation function, i got an error.

I may be able to work around that error, by using the new null provider technique from https://github.com/Jamie-BitFlight/terraform-null-label 

But yes, otherwise terraform automatically flattens it, however it has to have a value, empty lists cause error.